src/arch/arm/object/vcpu.c

```diff
@@ -161,54 +161,43 @@ void VGICMaintenance(void)
+
     flags = get_gic_vcpu_ctrl_misr();
 
     if (flags & VGIC_MISR_EOI) {
-        int irq_idx;
-        if (eisr0) {
-            irq_idx = ctzl(eisr0);
-        } else if (eisr1) {
-            irq_idx = ctzl(eisr1) + 32;
-        } else {
-            irq_idx = -1;
-        }
-
-        /* the hardware should never give us an invalid index, but we don't
-         * want to trust it that far */
-        if (irq_idx == -1  || irq_idx >= gic_vcpu_num_list_regs) {
-            current_fault = seL4_Fault_VGICMaintenance_new(0, 0);
-        } else {
-            virq_t virq = get_gic_vcpu_ctrl_lr(irq_idx);
-            switch (virq_get_virqType(virq)) {
-            case virq_virq_active:
-                virq = virq_virq_active_set_virqEOIIRQEN(virq, 0);
-                break;
-            case virq_virq_pending:
-                virq = virq_virq_pending_set_virqEOIIRQEN(virq, 0);
-                break;
-            case virq_virq_invalid:
-                virq = virq_virq_invalid_set_virqEOIIRQEN(virq, 0);
-                break;
+        uint64_t consumed_eisr = eisr;
+        while (consumed_eisr) {
+            int8_t irq_idx = ctzll(consumed_eisr);
+            /* the hardware should never give us an invalid index, but we don't
+             * want to trust it that far */
+            if (irq_idx < gic_vcpu_num_list_regs) {
+                virq_t virq;
+                /* Per spec definition, a vIRQ can only trigger an EOI IRQ iff its
+                 * (i) LR state is invalid, (ii) LR hw bit = 0, and (iii) LR eoi bit = 1.
+                 *
+                 * Since maintenance IRQ is level-sensitive, it is required to remove the
+                 * EOIed LRs, to avoid causing the maintenance IRQ again */
+                virq.words[0] = 0;
+                set_gic_vcpu_ctrl_lr(irq_idx, virq);
+
+                /* decodeVCPUInjectIRQ below checks the vgic.lr register,
+                 * so we should also sync the shadow data structure as well */
+                assert(ARCH_NODE_STATE(armHSCurVCPU) != NULL && ARCH_NODE_STATE(armHSVCPUActive));
+                if (ARCH_NODE_STATE(armHSCurVCPU) != NULL && ARCH_NODE_STATE(armHSVCPUActive)) {
+                    ARCH_NODE_STATE(armHSCurVCPU)->vgic.lr[irq_idx] = virq;
+                }
             }
-            set_gic_vcpu_ctrl_lr(irq_idx, virq);
-            /* decodeVCPUInjectIRQ below checks the vgic.lr register,
-             * so we should also sync the shadow data structure as well */
-            assert(ARCH_NODE_STATE(armHSCurVCPU) != NULL && ARCH_NODE_STATE(armHSVCPUActive));
-            if (ARCH_NODE_STATE(armHSCurVCPU) != NULL && ARCH_NODE_STATE(armHSVCPUActive)) {
-                ARCH_NODE_STATE(armHSCurVCPU)->vgic.lr[irq_idx] = virq;
-            } else {
-                /* FIXME This should not happen */
-            }
-            current_fault = seL4_Fault_VGICMaintenance_new(irq_idx, 1);
+            consumed_eisr &= ~(1ULL << irq_idx);
```

Comment:
Might be better to make no hard-coded assumption what type `uint64_t` actually is:
```suggestion
            /* clear bit at irq_idx */
            consumed_eisr &= ~((uint64_t)1 << irq_idx);
```

Reply:
Thanks, fixed.